File: app/src/main/java/com/maherhanna/cheeta/LegalMoves.java
package com.maherhanna.cheeta;

import java.util.ArrayList;
import java.util.HashMap;

public class LegalMoves {
    private ArrayList<Move> legalMoves;

    public LegalMoves(){
        legalMoves = new ArrayList<>();
    }

    public ArrayList<Integer> getLegalTargetsFor(int position)
    {
        ArrayList<Integer> targetSquares = new ArrayList<>();
        for(Move move: legalMoves){
            if(move.getFrom() == position){
                targetSquares.add(move.getTo());
            }
        }
        return targetSquares;
    }




    public boolean canMove(int from, int to){
        ArrayList<Integer> targetsForFrom = getLegalTargetsFor(from);
        if(targetsForFrom.isEmpty()) return false;
        if(targetsForFrom.contains(to)) return true;
        return false;
    }


    public int size(){
        return legalMoves.size();
    }
    public Move get(int index){
        return legalMoves.get(index);
    }

    public void add(Move newMove){
        legalMoves.add(newMove);
    }

    public void addAll(ArrayList<Move> moves){
        legalMoves.addAll(moves);
    }


    public Move getMove(Move basicMove) {
        Move legalMove = new Move(basicMove);
        for(Move move: legalMoves){
            if((move.getFrom() == basicMove.getFrom()) &&
                    (move.getTo() == basicMove.getTo())){
                legalMove = new Move(move);
                break;

            }
        }
        return legalMove;
    }
}

File: app/src/main/java/com/maherhanna/cheeta/LegalMoves.kt
package com.maherhanna.cheeta

class LegalMoves {
    private val legalMoves: ArrayList<Move>

    init {
        legalMoves = ArrayList()
    }

    fun getLegalTargetsFor(position: Int): ArrayList<Int> {
        val targetSquares = ArrayList<Int>()
        for (move in legalMoves) {
            if (move.from == position) {
                targetSquares.add(move.to)
            }
        }
        return targetSquares
    }

    fun canMove(from: Int, to: Int): Boolean {
        val targetsForFrom = getLegalTargetsFor(from)
        if (targetsForFrom.isEmpty()) return false
        return if (targetsForFrom.contains(to)) true else false
    }

    fun size(): Int {
        return legalMoves.size
    }

    operator fun get(index: Int): Move {
        return legalMoves[index]
    }

    fun add(newMove: Move) {
        legalMoves.add(newMove)
    }

    fun addAll(moves: ArrayList<Move>?) {
        legalMoves.addAll(moves!!)
    }

    fun getMove(basicMove: Move): Move {
        var legalMove = Move(basicMove)
        for (move in legalMoves) {
            if (move.from == basicMove.from && move.to == basicMove.to) {
                legalMove = Move(move)
                break
            }
        }
        return legalMove
    }

    fun removeNonTake() {
        val itr = legalMoves.iterator()
        while (itr.hasNext()) {
            val move = itr.next()
            if (!move.isTake) {
                itr.remove()
            }
        }
    }
}
